src/mrinufft/_array_compat.py

```diff
@@ -0,0 +1,379 @@
+
+        # get array module and device from non-self argument
+        xp = get_array_module(args[data_arg_idx])
+        device = _get_device(args[data_arg_idx])
+
+        # convert all to numpy
+        args = _to_numpy(*args)
+
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        return _to_interface(ret_, xp, device)
+
+    return wrapper
+
+
+def with_tensorflow(fun):
+    """Ensure the function works internally with tensorflow array."""
+    data_arg_idx = _ismethod(fun)
+
+    @wraps(fun)
+    def wrapper(*args, **kwargs):
+        args = _get_args(fun, args, kwargs)
+
+        # get array module from non-self argument
+        xp = get_array_module(args[data_arg_idx])
+        device = _get_device(args[data_arg_idx])
+
+        # convert all to tensorflow
+        args = _to_tensorflow(*args)
+
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        print(xp.__name__)
+        print(device)
+        return _to_interface(ret_, xp, device)
+
+    return wrapper
+
+
+def with_numpy_cupy(fun):
+    """Ensure the function works internally with numpy or cupy array."""
+    data_arg_idx = _ismethod(fun)
+
+    @wraps(fun)
+    def wrapper(*args, **kwargs):
+        args = _get_args(fun, args, kwargs)
+
+        # get array module from non-self argument
+        xp = get_array_module(args[data_arg_idx])
+
+        # convert all to cupy / numpy according to data arg device
+        args = _to_numpy_cupy(args, data_arg_idx)
+
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        return _to_interface(ret_, xp)
+
+    return wrapper
+
+
+def with_torch(fun):
+    """Ensure the function works internally with Torch."""
+    data_arg_idx = _ismethod(fun)
+
+    @wraps(fun)
+    def wrapper(*args, **kwargs):
+        args = _get_args(fun, args, kwargs)
+
+        # get array module from non-self argument
+        xp = get_array_module(args[data_arg_idx])
+        device = _get_device(args[data_arg_idx])
+
+        # convert all to tensorflow
+        args = _to_torch(*args)
+
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        return _to_interface(ret_, xp, device)
+
+    return wrapper
+
+
+def _ismethod(fun):
+    """Determine whether input fun is instance-/classmethod or not."""
+    first_arg = list(inspect.signature(fun).parameters)[0]
+    return first_arg in ["self", "cls"]
+
+
+def _get_device(input):
+    """Determine computational device from input array."""
+    try:
+        return input.device
+    except Exception:
+        return "cpu"
+
+
+def _get_args(func, args, kwargs):
+    """Convert input args/kwargs mix to a list of positional arguments.
+
+    This automatically fills missing kwargs with default values.
+    """
```

Comment:
The fact that it only returns positional arguments is tricky. if a wrapped function as **kwargs in its signature, the named parameters aspect is lost (which can be detrimental if their are handled further down in the wrapped function). I would advise to return args and kwargs separately. 

This would implies that the `_to_<array_module>` functions should process both args and kwargs

Reply:
At the beginning, I tried to handle both args and kwargs. The main problem I encountered was that if we pass only keyworded arguments, then things like `xp = get_array_module(args[data_arg_idx])` fail because there are no positional arguments left. The issue you are raising is indeed very valid - any suggestion on how to handle it? What about iterating over args and kwargs and picking the first arraylike argument as leading?